### ISSUE TITLE
Session data not being saved to redis after following redis session configuration instructions in documentation

### ISSUE THREAD

**Issue by sidm1983** [craftcms/cms#6796]
### Description
Hi, I have followed the instructions from the documentation on setting up redis as the session handler for Craft CMS (as well as researched online) and I am unable to get it working.

Basically, i have the session component section in config/app.php, but when I actually log in and then clear all my redis keys while logged in, I still remain logged in after that. I was expecting it to log me out.

It looks like PHP is still handling the session in the file system somehow, so I changed the session.save_handler setting in the php.ini file to `redis` and session.save_path to my redis connection string, but doing this throws a yii framework error saying that it couldn't find a session handler called `redis`. Was hoping someone could shed some light on this and point me in the right direction. Is there something missing from the craft cms documentation that I need to configure?

Also, for context, I am using docker for local development and my craft docker image is based on the official php:7.3-fpm image with nginx installed on top as a reverse proxy. Not sure if something in this setup is causing issues. The whole thing works really well, except for when I set that session.save_handler to redis. Looks like there is some more configuration I may be missing, but I am not sure where to look. I have already set up redis as the cache handler and that works well, but just having trouble correctly configuring sessions to be stored in redis as well.

Apologies if this is not a bug. I did ask about this on Discord but never got a response, so that's why I am asking here in case there is actually a bug preventing redis sessions from working.

Any help or guidance will be greatly appreciated 🙏

Thank you.

### Additional info

- Craft version: 3.5.8
- PHP version: 7.3.22
- Database driver & version: pgsql 11.9
- Plugins & versions:

**Comment by brandonkelly** [comment]
Have you defined a `redis` application component? See https://github.com/yiisoft/yii2-redis#configuration

**Comment by sidm1983** [comment]
@brandonkelly, yes, I have. Here is what my `config/app.php` file looks like:

```php
<?php
/**
 * Yii Application Config
 *
 * Edit this file at your own risk!
 *
 * The array returned by this file will get merged with
 * vendor/craftcms/cms/src/config/app.php and app.[web|console].php, when
 * Craft's bootstrap script is defining the configuration for the entire
 * application.
 *
 * You can define custom modules and system components, and even override the
 * built-in system components.
 *
 * If you want to modify the application config for *only* web requests or
 * *only* console requests, create an app.web.php or app.console.php file in
 * your config/ folder, alongside this one.
 */

use craft\helpers\App;

return [
    'id' => App::env('APP_ID') ?: 'CraftCMS',
    'modules' => [
        'my-module' => \modules\Module::class,
    ],
    'components' => [
        'redis' => [
            'class' => yii\redis\Connection::class,
            'hostname' => App::env('REDIS_HOST'),
            'port' => App::env('REDIS_PORT'),
            'password' => App::env('REDIS_PASSWORD'),
            'database' => 0,
            'useSSL' => (bool)App::env('REDIS_USE_SSL'),
        ],
        'cache' => [
            'class' => yii\redis\Cache::class,
            'defaultDuration' => 86400,
            'keyPrefix' => (App::env('APP_ID') ?: 'CraftCMS') . '__cache_',
        ],
        'session' => function() {
            // Get the default component config
            $config = App::sessionConfig();

            // Override the class to use Redis' session class
            $config['class'] = yii\redis\Session::class;

            // Instantiate and return it
            return Craft::createObject($config);
        },
        'mutex' => function() {
            $config = craft\helpers\App::mutexConfig();
            $config['isWindows'] = true;
            return Craft::createObject($config);
        },
    ],
    //'bootstrap' => ['my-module'],
];
```
The `cache` component configuration works by the way and I can see cache items created in redis and no longer being created in the filesystem.

Are there any other settings / configuration I should double-check?

Thank you.

**Comment by angrybrad** [comment]
If you query all of the values in Redis, then exclude any of them that have whatever `APP_ID` is set to or `CraftCMS` or `__cache_` for a prefix, do you have any that have a random alphanumeric key like `9eabfe68eceb16442a4a61526abf0603` with a serialized value that looks something like `a:2:{i:0;i:1599173022;i:1;N;}` ?

**Comment by sidm1983** [comment]
Hi @angrybrad,

Yes, I do have alphanumeric key and I know that is session related. Does that mean it is working? I tried deleting that alphanumeric key, but I am still logged in to the control panel even after deleting that key and it then creates a new alphanumeric key in redis for my session. Is this expected behaviour? It seems like the PHP session is still being managed in the filesystem or memory? Am I misunderstanding how the redis sessions work in this case? Does redis just store the session data, but it still gets hydrated into the filesystem / memory for handling by PHP?

Basically, I want to make sure that if I have multiple instances of my craft cms web server (i.e. highly available) then users don't get logged out when they hit different craft cms instances for each request.

Please let me know when you get a chance. Much appreciated. Thank you.

**Comment by brandonkelly** [comment]
Hey @sidm1983, it sounds like everything is working fine. Sorry I didn’t think of this earlier, but it’s totally expected that you would stay logged-in after clearing out Redis data, because user session data is stored in a browser cookie as a backup, in case they quit their browser or the session data gets garbage collected (as you are simulating when you delete all Redis data). The only things that would actually result in someone getting logged out are:

- They explicitly request to be logged out (e.g. visit `/logout`)
- Their session expires
- Their session’s row in the `sessions` table gets deleted

**Comment by sidm1983** [comment]
Thanks for the clarification @brandonkelly. I had mistakenly assumed that switching session handling to Redis meant that it was solely responsible for handling anything and everything to do with session. I did manage to test a highly available scenario on my local by having my craft cms container scaled to  3 instances and seeing whether I am automatically logged in to all 3 instances by only logging in to one of them. It all seems to be working as expected. So my last question is what is redis actually doing in this case and is it even needed in a multi-instance (HA) CMS setup? Based on your response above, it seems like the browser and the sessions table in the db is all that is needed to maintain a cms user's session. Would be great to get some clarification on the role of the Redis session data in this case.

Thank you.

**Comment by brandonkelly** [comment]
Well the user session info is just one thing that is stored in the session (whether that’s the PHP session or Redis). There are a few other things, like flash messages, the login return URL, authorization data, etc. Using a central session data storage is definitely needed if there are multiple web servers.

**Comment by sidm1983** [comment]
@brandonkelly thanks for the explanation. I am still quite new to Craft CMS and PHP, so some of this and the underlying architecture is still unclear to me. Not sure if there is any technical documentation I can read to familiarize myself with Craft's session architecture and concepts. For now, all good. It looks like it works as expected on my end. Thank you and @angrybrad for clarifying this and apologies for raising this as an issue when it really wasn't. ☺️🙏🏽
